base/strings/io.jl

```diff
@@ -341,15 +345,16 @@ julia> escape_string(string('\\u2135','\\0','0')) # \\0 would be ambiguous
 ## See also
 [`unescape_string`](@ref) for the reverse operation.
 """
-function escape_string(io::IO, s::AbstractString, esc="")
+function escape_string(io::IO, s::AbstractString, esc="", bsescaped = false)
```

Comment:
Should this be a keyword argument?

Reply:
I think it is better if it is a keyword argument because we may add more functionalities in the future. However, I was worried that this could decrease the performance of a function that is used in many places. Is it safe to change to keyword?

Reply:
yes, keyword arguments are fast

Reply:
OK! The version without the `io` I will change to

```julia
escape_string(s::AbstractString, esc=('\"',); bsescaped = false) =
                  sprint((io)->escape_string(s, esc; bsescaped = bsescaped), sizehint=lastindex(s))
```

Or is there a better way to do this? (`sprint` does not support kwargs of the function it calls AFAIK)

Reply:
Yes, I will do this. Benchmarks here show no change in performance at all.

Reply:
That looks fine to me. That said, I don't think `bescaped` is the best name if it's gonna be a keyword argument. Perhaps a better API would be to do something like `escape_string(...; keep='\\')`, though I'm not entirely happy with `keep` either, so alternative suggestions for that welcome.

Reply:
In this sense, I am definitely not a good guy to name keywords :) The idea is a boolean to tell that backslashes are already escaped.  Is `backslash_escaped` acceptable? I know Julia coding-style tells to avoid `_`, but otherwise it will be unreadable.

Reply:
Right, the problem I have with a boolean is that it's not clear whether a true value indicates that backslashes *should be* escaped or *are already* escaped. That's why I think adding essentially the inverse of the `esc` argument might be a better API since it would be unambiguous.

Reply:
Ah, I see, sorry :) Language problem here. Since we are keeping only backslashes, is it `keep_escaped_backslahes` good? `\\` is an escaped backslash AFAIK. It is big, but very informative.

Reply:
I think that or `escape_backslash=false` would be fine, but it's both a bit long and raises generality questions (e.g. are we gonna add `escape_newline`, `escape_tab`, etc next when people ask for them or should we just do the general thing now).

Reply:
Ok, so I propose the following: let's add a keyword (or another argument) called `keep`, as you proposed and what is used in `unescape_string`. `keep` is a string. Every character inside `keep` will be left untouched. Furthermore, `esc` has precedence. Thus, if we have `esc = "\\", keep = "\\"`, then the backslashes will be escaped.

If this is acceptable, should `keep` be an argument or keyword? I am asking this because `keep` is an argument in `unscape_string`.

Reply:
I think it should be a keyword argument. Also I think defaulting it to empty tuple makes more sense, so the check can be specialized away by default. Others should chime in on the naming question also though.

Reply:
I will update everything considering `keep`, then we wait for some suggestions. And yes, I will make it a tuple by default (`keep = ()`) just as in `unescape_string`.

Side question, please, answer only if you have time: why selecting it as a tuple by default helps the specialization compared to an empty string?

Reply:
Empty tuples are known to be empty at compile time, but empty strings are not.

Reply:
Done!